### PR TITLE
Query: Fixes ORDER BY undefined (and mixed type primitives) continuation token support

### PR DESCRIPTION
**Pull request by bchong95**
# Query: Fixes ORDER BY undefined continuation token support

## Description

For ORDER BY queries we only store the continuation token for the partition we are resuming on. For the other partitions (left of target and right of target) we generate a filter that mimics the resume term. The problem here is that we didn't account for the undefined case when writing our visitor pattern. The following code results in a NullReferenceException:

`orderByItem.Accept(cosmosElementToQueryLiteral);`

Since `orderByItem` is null. 

One solution to avoid the NRE is to use the `undefined` literal in the filter. Unfortunately this would lead to correctness issues, since comparing against undefined like so:

`c.item > undefined` 

is also undefined. leading to the entire query returning no results. 

So the solution is to use the `IS_DEFINED` system function to remove the undefined fields when needed.

### CODE REVIEW

File: Microsoft.Azure.Cosmos/src/Query/Core/Pipeline/CrossPartition/OrderBy/OrderByCrossPartitionQueryPipelineStage.cs [hunk 1]
```diff
@@ -816,15 +816,48 @@ private static void AppendToBuilders((StringBuilder leftFilter, StringBuilder ta
                 (OrderByColumn orderByColumn, CosmosElement orderByItem) = columnAndItems.Span[0];
                 (string expression, SortOrder sortOrder) = (orderByColumn.Expression, orderByColumn.SortOrder);
 
-                StringBuilder sb = new StringBuilder();
-                CosmosElementToQueryLiteral cosmosElementToQueryLiteral = new CosmosElementToQueryLiteral(sb);
-                orderByItem.Accept(cosmosElementToQueryLiteral);
+                if (orderByItem == default)
```
Comment (j82w) [inline Microsoft.Azure.Cosmos/src/Query/Core/Pipeline/CrossPartition/OrderBy/OrderByCrossPartitionQueryPipelineStage.cs:819]: Please add the PR that introduced the regression into the description so we can note it in the changelog.

Comment (j82w) [reply Microsoft.Azure.Cosmos/src/Query/Core/Pipeline/CrossPartition/OrderBy/OrderByCrossPartitionQueryPipelineStage.cs:819]: Is this issue by any chance related to this fix? https://github.com/Azure/azure-cosmos-dotnet-v3/issues/1364

File: Microsoft.Azure.Cosmos/src/Query/Core/Pipeline/CrossPartition/OrderBy/OrderByCrossPartitionQueryPipelineStage.cs [hunk 2]
```diff
@@ -1058,5 +1160,130 @@ public void SetCancellationToken(CancellationToken cancellationToken)
...
+            public string LessThanOrEqualTo { get; }
+            public string EqualTo { get; }
+            public string GreaterThan { get; }
+            public string GreaterThanOrEqualTo { get; }
+
+            public static ComparisionWithUndefinedFilters Create(string expression)
+            {
+                return new ComparisionWithUndefinedFilters(
+                    lessThan: "false",
+                    lessThanOrEqualTo: $"not IS_DEFINED({expression})",
```
Comment (sboshra) [inline Microsoft.Azure.Cosmos/src/Query/Core/Pipeline/CrossPartition/OrderBy/OrderByCrossPartitionQueryPipelineStage.cs:1282]: NOT

File: Microsoft.Azure.Cosmos/src/Query/Core/Pipeline/CrossPartition/OrderBy/OrderByCrossPartitionQueryPipelineStage.cs [hunk 3]
```diff
@@ -1058,5 +1160,130 @@ public void SetCancellationToken(CancellationToken cancellationToken)
...
+                IsSystemFunctions.Undefined,
+                IsSystemFunctions.Null,
+                IsSystemFunctions.Boolean,
+                IsSystemFunctions.Number,
+                IsSystemFunctions.String,
+                IsSystemFunctions.Array,
+                IsSystemFunctions.Object,
+            };
+
+            private static class Indexes
```
Comment (sboshra) [inline Microsoft.Azure.Cosmos/src/Query/Core/Pipeline/CrossPartition/OrderBy/OrderByCrossPartitionQueryPipelineStage.cs:1190]: SortOrder

File: Microsoft.Azure.Cosmos/src/Query/Core/Pipeline/CrossPartition/OrderBy/OrderByCrossPartitionQueryPipelineStage.cs [hunk 4]
```diff
@@ -1058,5 +1160,130 @@ public void SetCancellationToken(CancellationToken cancellationToken)
...
+                this.GreaterThanOrEqualTo = greaterThanOrEqualTo ?? throw new ArgumentNullException(nameof(greaterThanOrEqualTo));
+            }
+
+            public string LessThan { get; }
+            public string LessThanOrEqualTo { get; }
+            public string EqualTo { get; }
+            public string GreaterThan { get; }
+            public string GreaterThanOrEqualTo { get; }
+
+            public static ComparisionWithUndefinedFilters Create(string expression)
```
Comment (sboshra) [inline Microsoft.Azure.Cosmos/src/Query/Core/Pipeline/CrossPartition/OrderBy/OrderByCrossPartitionQueryPipelineStage.cs:1278]: constructor

**Review by sboshra — APPROVED** [review]

File: Microsoft.Azure.Cosmos/src/Query/Core/Pipeline/CrossPartition/OrderBy/OrderByCrossPartitionQueryPipelineStage.cs (hunk 1 above)
Comment (bchong95) [reply Microsoft.Azure.Cosmos/src/Query/Core/Pipeline/CrossPartition/OrderBy/OrderByCrossPartitionQueryPipelineStage.cs]: It's a day zero bug. Unrelated to that fix.

File: Microsoft.Azure.Cosmos/src/Query/Core/Pipeline/CrossPartition/OrderBy/OrderByCrossPartitionQueryPipelineStage.cs (hunk 2 above)
Comment (bchong95) [reply Microsoft.Azure.Cosmos/src/Query/Core/Pipeline/CrossPartition/OrderBy/OrderByCrossPartitionQueryPipelineStage.cs:1282]: done.

File: Microsoft.Azure.Cosmos/src/Query/Core/Pipeline/CrossPartition/OrderBy/OrderByCrossPartitionQueryPipelineStage.cs (hunk 4 above)
Comment (bchong95) [reply Microsoft.Azure.Cosmos/src/Query/Core/Pipeline/CrossPartition/OrderBy/OrderByCrossPartitionQueryPipelineStage.cs:1278]: done.

File: Microsoft.Azure.Cosmos/src/Query/Core/Pipeline/CrossPartition/OrderBy/OrderByCrossPartitionQueryPipelineStage.cs (hunk 3 above)
Comment (bchong95) [reply Microsoft.Azure.Cosmos/src/Query/Core/Pipeline/CrossPartition/OrderBy/OrderByCrossPartitionQueryPipelineStage.cs:1190]: done.

**Review by sboshra — APPROVED** [review]

**Pull request merged**